presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryProvider.java

```diff
@@ -0,0 +1,79 @@
+ */
+package com.facebook.presto.orc.reader;
+
+import com.facebook.presto.orc.OrcCorruptionException;
+import com.facebook.presto.orc.OrcLocalMemoryContext;
+import com.facebook.presto.orc.StreamDescriptor;
+import com.facebook.presto.orc.stream.InputStreamSource;
+import com.facebook.presto.orc.stream.InputStreamSources;
+import com.facebook.presto.orc.stream.LongInputStream;
+
+import java.io.IOException;
+
+import static com.facebook.presto.orc.metadata.Stream.StreamKind.DICTIONARY_DATA;
+import static io.airlift.slice.SizeOf.sizeOf;
+import static java.util.Objects.requireNonNull;
+
+public class LongDictionaryProvider
+{
+    private final InputStreamSources dictionaryStreamSources;
+    private final OrcLocalMemoryContext memoryUsage;
+    private long totalBytes;
+
+    public LongDictionaryProvider(InputStreamSources dictionaryStreamSources, OrcLocalMemoryContext memoryUsage)
+    {
+        this.dictionaryStreamSources = requireNonNull(dictionaryStreamSources, "dictionaryStreamSources is null");
+        this.memoryUsage = requireNonNull(memoryUsage, "memoryUsage is null");
+        this.totalBytes = 0;
+    }
+
+    /**
+     * Loads a dictionary from a stream and attempts to reuse the dictionary buffer passed in.
+     *
+     * @param streamDescriptor descriptor indicating node and sequence of the stream reader
+     * the dictionary is associated with.
+     * @param dictionary dictionary buffer the method attempts to fill.
+     * @param items number of items expected in the dictionary.
+     * @return The final dictionary buffer object. Different from the input dictionary buffer if
+     * the input dictionary buffer is expanded.
+     * @throws IOException
+     */
+    public long[] getDictionary(StreamDescriptor streamDescriptor, long[] dictionary, int items)
+            throws IOException
+    {
+        InputStreamSource<LongInputStream> dictionaryDataStream = dictionaryStreamSources.getInputStreamSource(streamDescriptor, DICTIONARY_DATA, LongInputStream.class);
+        return loadDictionary(streamDescriptor, requireNonNull(dictionaryDataStream), dictionary, items);
+    }
+
+    private long[] loadDictionary(StreamDescriptor streamDescriptor, InputStreamSource<LongInputStream> dictionaryDataStream, long[] dictionaryBuffer, int items)
+            throws IOException
+    {
+        // We construct and use the input stream exactly once per stream descriptor per stripe, so we don't
+        // really need to cache it.
+        LongInputStream inputStream = dictionaryDataStream.openStream();
+        if (inputStream == null) {
+            throw new OrcCorruptionException(streamDescriptor.getOrcDataSourceId(), "Dictionary is not empty but data stream is not present for %s", streamDescriptor);
+        }
+
+        if (dictionaryBuffer == null || dictionaryBuffer.length < items) {
+            dictionaryBuffer = new long[items];
+            totalBytes += sizeOf(dictionaryBuffer);
+            memoryUsage.setBytes(totalBytes);
```

Comment:
New accounting in dictionary provider. Side effect is that now selective stream reader case also records dictionary memory usage, but I think that is for the better.